File: library/swift/src/EngineImpl.swift
@_implementationOnly import EnvoyEngine
import Foundation

/// Envoy Mobile Engine implementation.
@objcMembers
final class EngineImpl: NSObject {
  private let engine: EnvoyEngine
  private let statsClientImpl: StatsClientImpl
  private let streamClientImpl: StreamClientImpl

  private enum ConfigurationType {
    case yaml(String)
    case typed(EnvoyConfiguration)
  }

  private init(configType: ConfigurationType, logLevel: LogLevel, engine: EnvoyEngine,
               onSetupComplete: (() -> Void)?)
  {
    self.engine = engine
    self.statsClientImpl = StatsClientImpl(engine: engine)
    self.streamClientImpl = StreamClientImpl(engine: engine)
    super.init()

    switch configType {
    case .yaml(let configYAML):
      self.engine.run(withConfigYAML: configYAML, logLevel: logLevel.stringValue,
                      onSetupComplete: onSetupComplete)
    case .typed(let config):
      self.engine.run(withConfig: config, logLevel: logLevel.stringValue,
                      onSetupComplete: onSetupComplete)
    }
  }

  /// Initialize a new Envoy instance using a typed configuration.
  ///
  /// - parameter config:          Configuration to use for starting Envoy.
  /// - parameter logLevel:        Log level to use for this instance.
  /// - parameter engine:          The underlying engine to use for starting Envoy.
  /// - parameter onSetupComplete: Closure called when the engine finishes its async
  ///                              initialization/startup.
  convenience init(config: EnvoyConfiguration, logLevel: LogLevel = .info, engine: EnvoyEngine,
                   onSetupComplete: (() -> Void)?)
  {
    self.init(configType: .typed(config), logLevel: logLevel, engine: engine,
              onSetupComplete: onSetupComplete)
  }

  /// Initialize a new Envoy instance using a string configuration.
  ///
  /// - parameter configYAML:      Configuration yaml to use for starting Envoy.
  /// - parameter logLevel:        Log level to use for this instance.
  /// - parameter engine:          The underlying engine to use for starting Envoy.
  /// - parameter onSetupComplete: Closure called when the engine finishes its async
  ///                              initialization/startup.
  convenience init(configYAML: String, logLevel: LogLevel = .info, engine: EnvoyEngine,
                   onSetupComplete: (() -> Void)?)
  {
    self.init(configType: .yaml(configYAML), logLevel: logLevel, engine: engine,
              onSetupComplete: onSetupComplete)
  }
}

extension EngineImpl: Engine {
  func streamClient() -> StreamClient {
    return self.streamClientImpl
  }

  func statsClient() -> StatsClient {
    return self.statsClientImpl
  }
}

File: library/kotlin/io/envoyproxy/envoymobile/EngineImpl.kt
package io.envoyproxy.envoymobile

import io.envoyproxy.envoymobile.engine.EnvoyConfiguration
import io.envoyproxy.envoymobile.engine.EnvoyEngine

/**
 * An implementation of {@link Engine}.
 */
class EngineImpl constructor(
  internal val envoyEngine: EnvoyEngine,
  internal val envoyConfiguration: EnvoyConfiguration,
  internal val configurationYAML: String?,
  internal val logLevel: LogLevel,
  internal val componentLogLevel: String
) : Engine {

  private val streamClient: StreamClient
  private val pulseClient: PulseClient

  constructor(
    envoyEngine: EnvoyEngine,
    envoyConfiguration: EnvoyConfiguration,
    logLevel: LogLevel = LogLevel.INFO,
    componentLogLevel: String
  ) : this(envoyEngine, envoyConfiguration, null, logLevel, componentLogLevel)

  init {
    streamClient = StreamClientImpl(envoyEngine)
    pulseClient = PulseClientImpl(envoyEngine)
    if (configurationYAML != null) {
      envoyEngine.runWithTemplate(configurationYAML, envoyConfiguration, logLevel.level, componentLogLevel)
    } else {
      envoyEngine.runWithConfig(envoyConfiguration, logLevel.level, componentLogLevel)
    }
  }

  override fun streamClient(): StreamClient {
    return streamClient
  }

  override fun pulseClient(): PulseClient {
    return pulseClient
  }

  override fun terminate() {
    envoyEngine.terminate()
  }

  override fun flushStats() {
    envoyEngine.flushStats()
  }

  override fun dumpStats(): String {
    return envoyEngine.dumpStats()
  }

  override fun drainConnections() {
    envoyEngine.drainConnections()
  }
}
